brew_view/authorization.py

```diff
@@ -0,0 +1,157 @@
+    USER_READ = 'bg-user-read'
+    USER_UPDATE = 'bg-user-update'
+    USER_DELETE = 'bg-user-delete'
+
+
+def anonymous_user():
+    """Get a Principal representing an anonymous user
+
+    Returns:
+        BrewtilsPrincipal
+    """
+    return BrewtilsPrincipal(roles=['bg-anonymous'],
+                             permissions=brew_view.anonymous_permissions)
+
+
+def authenticated(method=None, permissions=None):
+    """Decorate methods with this to require various permissions"""
+
+    if method is None:
+        return functools.partial(authenticated, permissions=permissions)
+
+    # Convert to strings for easier comparison
+    permission_strings = set(p.value for p in permissions)
+
+    @functools.wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if not has_permission(self.current_user, permission_strings):
+            raise RequestForbidden('Action requires permission %s' %
+                                   permissions[0].value)
+
+        return method(self, *args, **kwargs)
+
+    return wrapper
+
+
+def has_permission(principal, required_permissions):
+    """Determine if a principal has access to a resource
+
+    Args:
+        principal: the principal to test
+        required_permissions: set of strings
+
+    Returns:
+        bool yes or no
+    """
+    if Permissions.ALL.value in principal.permissions:
+        return True
+
+    return bool(required_permissions.intersection(principal.permissions))
+
+
+def coalesce_permissions(role_list):
+    """Determine permissions"""
+
+    if not role_list:
+        return set(), set()
+
+    aggregate_roles = set()
+    aggregate_perms = set()
+
+    for role in role_list:
+        aggregate_roles.add(role.name)
+        aggregate_perms |= set(role.permissions)
+
+        nested_roles, nested_perms = coalesce_permissions(role.roles)
+        aggregate_roles |= nested_roles
+        aggregate_perms |= nested_perms
+
+    return aggregate_roles, aggregate_perms
+
+
+def basic_auth(auth_header):
+    """Determine if a basic authorization header is valid
+
+    Args:
+        auth_header: The Authorization header. Should start with 'Basic '.
+
+    Returns:
+        Brewtils principal if auth_header is valid, None otherwise
+    """
+    auth_decoded = base64.b64decode(auth_header[6:]).decode()
+    username, password = auth_decoded.split(':')
+
+    try:
+        principal = Principal.objects.get(username=username)
+
+        if custom_app_context.verify(password, principal.hash):
+            return principal
+    except DoesNotExist:
+        # Don't handle this differently to prevent an attacker from being able
+        #  to enumerate a list of user names
+        pass
+
+    return None
+
+
+def bearer_auth(auth_header):
+    """Determine if a bearer authorization header is valid
+
+    Args:
+        auth_header: The Authorization header. Should start with 'Bearer '.
+
+    Returns:
+        Brewtils principal if auth_header is valid, None otherwise
```

Comment:
Does this actually return `None` ever? Also, why are we using a `BrewtilsPrincipal` here instead of a `bg-utils` model?

Reply:
Reworked this a little to make more sense. We return a brewtils model here because the bg-utils flavor doesn't have permissions, only roles.